Classes/Promise.swift

```diff
@@ -0,0 +1,269 @@
+                self.onCancel = wrappedBlock
+            }
+        }
+    }
+    
+    func cancel() {
+        cancelled = true
+        OSMemoryBarrier()
+        PromiseCancelToken.queue.async {
+            self.onCancel?()
+            self.onCancel = nil
+        }
+    }
+}
+
+public class Promise<ValueType> : PromiseCancelToken {
+    
+    var fulfilledObject: PromiseResult<ValueType>?
+    
+    public class PromiseBlock {
+        var done : ((ValueType) -> Void)?
+        var observed : ((ValueType) -> Void)?
+        var rejected : ((Error) -> Void)?
+        var always : (() -> Void)?
+        var queue : DispatchQueue?
+        let cancellationToken : PromiseCancelToken
+        
+        init(cancellationToken: PromiseCancelToken) {
+            self.cancellationToken = cancellationToken
+        }
+        var shouldKeepPromise: Bool {
+            get {
+                return done != nil || always != nil || rejected != nil
+            }
+        }
+        
+        private func internalCall(with object: PromiseResult<ValueType>) {
+            if !self.cancellationToken.cancelled {
+                if case let .failure(error) = object {
+                    rejected?(error)
+                } else if case let .success(value) = object {
+                    done?(value)
+                    observed?(value)
+                }
+                always?()
+            }
+        }
+        
+        func call(with object: PromiseResult<ValueType>) {
+            if let queue = queue {
+                queue.async {
+                    self.internalCall(with: object)
+                }
+            } else {
+                internalCall(with: object)
+            }
+        }
+    }
+    
+    @discardableResult func then(_ onFulfilled: ((ValueType) -> Void)? = nil,
+                                 observed: ((ValueType) -> Void)? = nil,
+                                 rejected : ((Error) -> Void)? = nil,
+                                 always : (() -> Void)? = nil,
+                                 queue : DispatchQueue) -> PromiseCancelToken {
+        let cancellationToken = PromiseCancelToken()
+        let blocks = PromiseBlock(cancellationToken: cancellationToken)
+        blocks.queue = queue
+        blocks.done = onFulfilled
+        blocks.observed = observed
+        blocks.rejected = rejected
+        blocks.always = always
+        
+        cancellationToken.cancelled({ [weak self, unowned blocks] in
+            Promise.queue.async {
+                if let `self` = self {                    
+                    blocks.done = nil
+                    blocks.observed = nil
+                    blocks.rejected = nil
+                    blocks.always = nil
+                    
+                    var strongCount = false
+                    for block in `self`.blocks {
+                        if block.shouldKeepPromise {
+                            strongCount = true
+                            break
```

Comment:
from my point of view would be more readable one line conditional
`for block in blocks where block.shouldKeepPromise`

Reply:
More swifty would be 
```
if let strongCount = blocks.contains(where: { $0.shouldKeepPromise }) {
     `self`.cancel()
}
```